### PR TITLE
Make static page prefetch apply on navigation in app dir

### PR DESCRIPTION
**Pull request by timneutkens**
The initial prefetching implementation was based on the response returning below the common layout, however when static generation was added the RSC payload for these changed to only include the router tree patch and the fully rendered page.

Currently the flow for prefetch is this:

- `router.prefetch`
- Fetch RSC payload
- Dispatch `ACTION_PREFETCH` with RSC payload. Note: the fetch is intentionally not in the reducer as the fetch happens during an event or effect, e.g. hover or `useEffect`
- Reducer handles `ACTION_PREFETCH`, creates the router tree, applies the `subTreeData` to the cache append-only. Saves the new router tree in the `prefetchCache` part of the reducer state.
- Then, on navigation, the `subTreeData` should be already available so there's no additional apply. The router state is applied based on the `prefetchCache`.

This approach is fine when the RSC payload returned never overrides the `subTreeData` of an already existing node, however that does happen in case of static pages because the subTreeData is for the root cache node instead of a deeper node.

The new flow for prefetching:

- `router.prefetch`
- Fetch RSC payload
- Dispatch `ACTION_PREFETCH` with RSC payload
- Reducer handles `ACTION_PREFETCH`, creates the router tree, Saves the new router tree and `subTreeData` in the `prefetchCache` part of the reducer state.
- Then, on navigation, the `subTreeData` and router state are  applied based on the `prefetchCache`.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm build && pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| buildDuration | 46.3s | 47.2s | ⚠️ +886ms |
| buildDurationCached | 12.6s | 12.9s | ⚠️ +340ms |
| nodeModulesSize | 95 MB | 95 MB | ✓ |
| nextStartRea..uration (ms) | 343ms | 333ms | -10ms |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| 152-HASH.js gzip | 64.5 kB | 64.5 kB | ✓ |
| 53.HASH.js gzip | 182 B | 182 B | ✓ |
| main-app-HASH.js gzip | 205 B | 205 B | ✓ |
| main-HASH.js gzip | 79.2 kB | 79.2 kB | ✓ |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 146 kB | 146 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| _app-HASH.js gzip | 192 B | 192 B | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 484 B | 484 B | ✓ |
| css-HASH.js gzip | 804 B | 804 B | ✓ |
| dynamic-HASH.js gzip | 2.29 kB | 2.29 kB | ✓ |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 827 B | 827 B | ✓ |
| hooks-HASH.js gzip | 848 B | 848 B | ✓ |
| image-HASH.js gzip | 4.3 kB | 4.3 kB | ✓ |
| index-HASH.js gzip | 252 B | 252 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | ✓ |
| routerDirect..HASH.js gzip | 782 B | 782 B | ✓ |
| script-HASH.js gzip | 857 B | 857 B | ✓ |
| withRouter-HASH.js gzip | 781 B | 781 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.6 kB | 15.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 481 B | ✓ |
| Overall change | 481 B | 481 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| index.html gzip | 489 B | 489 B | ✓ |
| link.html gzip | 504 B | 504 B | ✓ |
| withRouter.html gzip | 485 B | 485 B | ✓ |
| Overall change | 1.48 kB | 1.48 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| edge-ssr.js gzip | 110 kB | 110 kB | ✓ |
| page.js gzip | 98.4 kB | 98.4 kB | -1 B |
| Overall change | 209 kB | 209 kB | -1 B |

</details>

<details>
<summary><strong>Middleware size</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/prefetch-static-page | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 585 B | 583 B | -2 B |
| middleware-r..fest.js gzip | 145 B | 145 B | ✓ |
| middleware.js gzip | 27 kB | 27 kB | ✓ |
| edge-runtime..pack.js gzip | 1.83 kB | 1.83 kB | ✓ |
| Overall change | 29.6 kB | 29.6 kB | -2 B |

</details>

</details>
Commit: 6e1e6808d1b2a472befd79f9ac561d345b997e38

**Review by sebmarkbage — COMMENTED** [review]
I think this in general makes sense. My main concern would be that this can override existing cached data that might actually be newer than what is in the prefetch cache.

When they're inserted into the regular cache, they naturally fall away when we're navigating away from various subtrees or issue refreshes.

We'd need some way to purge them from the prefetch cache.

**Comment by timneutkens** [comment]
Potentially we can remove the entry from the prefetch cache when the navigation is applied, because of the soft navigation heuristic we fill the cache the first time the page is rendered and then remove the prefetch entry. This would cause it to be prefetched again later. However this does not fix the existing cache data being overwritten because of the way that the subTreeData is returned for the root and we need to render to fill the cache. The way around this would be to split the RSC response so that the layout / page levels are separate payloads in the response 🤔

File: test/e2e/app-dir/app-prefetch/app/page.js [hunk 1]
```diff
@@ -5,6 +5,9 @@ export default function HomePage() {
       <Link href="/dashboard" id="to-dashboard">
         To Dashboard
       </Link>
+      <Link href="/static-page" id="to-static-page">
```
Comment (hanneslund) [inline test/e2e/app-dir/app-prefetch/app/page.js:8]: This is added but not used in any test?

File: test/e2e/app-dir/app-prefetch/app/static-page/page.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export default async function Page() {
```
Comment (hanneslund) [inline test/e2e/app-dir/app-prefetch/app/static-page/page.js:1]: And this path as well?

**Review by ijjk — APPROVED** [review]

**Pull request merged**